// datocms-client/lib/ApiException.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ApiException;

function ApiException(response, body) {
  if ('captureStackTrace' in Error) {
    Error.captureStackTrace(this, ApiException);
  } else {
    this.stack = new Error().stack;
  }

  if (response) {
    if (response.status < 500) {
      var error = body.data[0];
      var details = JSON.stringify(error.attributes.details);
      this.message = "".concat(response.status, " ").concat(error.attributes.code, " (details: ").concat(details, ")");
    } else {
      this.message = "".concat(response.status, " ").concat(response.statusText);
    }

    this.body = body;
    this.headers = response.headers;
    this.statusCode = response.status;
    this.statusText = response.statusText;
  } else {
    this.message = 'Misconfigured exception';
  }
}

ApiException.prototype = Object.create(Error.prototype);
ApiException.prototype.name = 'ApiException';
ApiException.prototype.constructor = ApiException;

ApiException.prototype.errorWithCode = function errorWithCode(code) {
  if (!this.body || !(this.body.data instanceof Array)) {
    return null;
  }

  return this.body.data.find(function (error) {
    return error.attributes.code === code;
  });
};
