// datocms-client/lib/upload/adapters/nodeLocal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = nodeLocal;

var _axios = _interopRequireDefault(require("axios"));

var _path = _interopRequireDefault(require("path"));

var _fs = _interopRequireDefault(require("fs"));

var _mimeTypes = _interopRequireDefault(require("mime-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function uploadToS3(url, filePath) {
  return (0, _axios["default"])({
    url: url,
    method: 'put',
    headers: {
      'content-type': _mimeTypes["default"].lookup(filePath)
    },
    data: _fs["default"].readFileSync(filePath),
    maxContentLength: 1000000000
  });
}

function nodeLocal(client, filePath) {
  return client.uploadRequest.create({
    filename: _path["default"].basename(filePath)
  }).then(function (_ref) {
    var id = _ref.id,
        url = _ref.url;
    return uploadToS3(url, filePath).then(function () {
      return id;
    });
  });
}
